tutorials/rag/agentic_rag.py

```diff
@@ -0,0 +1,280 @@
+            "parameters": {
+                "type": "object",
+                "required": ["question"],
+                "properties": {
+                    "question": {"type": "string", "description": "Details of the clarification you would like from the user in order to help answer the question."},
+                },
+                "additionalProperties": False,
+            },
+        },
+    },
+    {
+        "type": "function",
+        "function": {
+            "name": "provide_answer",
+            "description": "Submits the answer to the user.",
+            "parameters": {
+                "type": "object",
+                "required": ["answer", "reasoning", "citation"],
+                "properties": {
+                    "answer": {"type": "string", "description": "The final synthesized answer."},
+                    "reasoning": {"type": "string", "description": "The reasoning for your answer."},
+                    "citation": {"type": "string", "description": "The citation to the knowledge base entry you used when answering."},
+                },
+                "additionalProperties": False,
+            },
+        },
+    }
+]
+
+
+# Implement the mock tools themselves
+def retrieve_knowledge(query: str) -> str:
+    """Retrieve most relevant document from the vector db (Chroma) for the question."""
+    print("RETRIEVER:\n", query)
+    response = collection.query(query_texts=[query], n_results=1)
+    retrieved_doc = response["documents"][0][0]
+    return retrieved_doc
+
+
+def get_clarification(question: str) -> str:
+    """Ask the user for clarifying information."""
+    # TODO: Implement this with your own logic for sending and receiving messages from the user
+    print("AI:\n", question)
+    return input("User:\n")
+
+
+def provide_answer(answer: str, reasoning: str, citation: str) -> str:
+    """Provide the answer to the user."""
+    # TODO: Implement this with your own logic for sending the final response to the user
+    print(
+        f"\n\nAnswer: {answer}\nReasoning: {reasoning}\nCitation: {citation}"
+    )
+
+
+def call_model(inputs: dict[str, str], messages: list[dict], trace_id: str) -> dict:
+    """Calls the model with the provided inputs and messages."""
+
+    # Populate the Prompt template
+    start_time = datetime.now()
+    populated_template = hl.prompts.populate_template(
+        template=PROMPT_TEMPLATE,
+        inputs=inputs
+    )
+    # Call OpenAI to get response - note you can also instead manage your prompts on Humanloop and use our proxy:
+    # https://humanloop.com/docs/v5/guides/prompts/call-prompt
+    chat_completion = openai.chat.completions.create(
+        model=MODEL,
+        messages=populated_template + messages,
+        # include the tool schemas so the model can make a call
+        tools=TOOL_SCHEMAS,
+    )
+    # Log to Humanloop
+    hl.prompts.log(
+        path=f"{APP_NAME}/medqa-answer",
```

Comment:
prompt name should be sentence case (and flow name below)

(tool name is "Knowledge base" which is sentence case)